[video][docs] Document the ways of listening to events (#32242)
expo/expo: docs/pages/versions/unversioned/sdk/video.mdx

</SnackInline>

> ## Receiving events

Reviewer: I'm wondering what would be better here – putting this section before the API section or after? I feel like, as a docs reader, I would like to read the API first and then some guides, more advanced examples etc.

Author: I feel like in most libraries we keep the usage examples at the top, so I guess the current way would be more consistent

The changes in properties of the [`VideoPlayer`](#videoplayer) do not update the React state, therefore in order to display the information about the current state of the VideoPlayer it is necessary to listen to the [events](#videoplayerevents) it emits. There are a few ways to listen to events:

> - `useEvent` hook - this hook creates a listener that will return a stateful value that can be used in a component. It also cleans up automatically when the component unmounts.

Reviewer: We could add a link to the `expo` package docs for more details.

```tsx
import { useEvent } from 'expo';
import { useVideoPlayer } from 'expo-video';

export default function ListenToStatus() {
  const player = useVideoPlayer(videoSource);
  const { status, error } = useEvent(player, 'statusChange', { status: player.status });

  // Rest of the component...
}
```

> - `useEventListener` hook - built around the `Player.addEventListener` and `Player.removeEventListener` methods, creates an event listener with automatic cleanup.

Reviewer: ```suggestion
- `useEventListener` hook - built around the `player.addListener` and `player.removeListener` methods, creates an event listener with automatic cleanup.
```

```tsx
import { useEventListener } from 'expo';
import { PlayerError, useVideoPlayer } from 'expo-video';
import { useState } from 'react';

export default function ListenToStatus() {
  const player = useVideoPlayer(videoSource);
  const [playerStatus, setPlayerStatus] = useState(player.status);
  const [playerError, setPlayerError] = useState<PlayerError | null>(null);

  useEventListener(player, 'statusChange', ({ status, error }) => {
    setPlayerStatus(status);
    setPlayerError(error);
    console.log('Player status changed: ', status);
  });

  // Rest of the component...
}
```

> - `Player.addEventListener` method - this is the most flexible way to listen to events, but requires manual cleanup and more boilerplate code.

Reviewer: ```suggestion
- `player.addListener` and `player.removeListener` methods - this is the most flexible way to listen to events, but requires manual cleanup and more boilerplate code.
```

```tsx
import { useEffect, useState } from 'react';
import { VideoPlayerEvents, useVideoPlayer, PlayerError } from 'expo-video';

function ListenToStatus() {
  const player = useVideoPlayer(videoSource);
  const [playerStatus, setPlayerStatus] = useState(player.status);
  const [playerError, setPlayerError] = useState<PlayerError | null>(null);

  useEffect(() => {

>     const listener: VideoPlayerEvents['statusChange'] = ({ status, error }) => {
>       setPlayerStatus(status);
>       setPlayerError(error);
>       console.log('Player status changed: ', status);
>     };
>     player.addListener('statusChange', listener);

Reviewer: It might be better to just inline the listener, then you don't need to import `VideoPlayerEvents` and the IntelliSense works nicely as you type – once you finish typing the event name, you'll get all the info about listener parameters.

```suggestion
    player.addListener('statusChange', ({ status, error }) => {
      setPlayerStatus(status);
      setPlayerError(error);
      console.log('Player status changed: ', status);
    });
```

</SnackInline>

### Receiving events

> The changes in properties of the [`VideoPlayer`](#videoplayer) do not update the React state, therefore in order to display the information about the current state of the VideoPlayer it is necessary to listen to the [events](#videoplayerevents) it emits.

Reviewer 2: ```suggestion
The changes in properties of the [`VideoPlayer`](#videoplayer) do not update the React state. Therefore, to display the information about the current state of the `VideoPlayer`, it is necessary to listen to the [events](#videoplayerevents) it emits.
```

The event system is based on the [`EventEmitter`](../sdk/expo/#eventemitter) class and [hooks](../sdk/expo/#hooks) from the [`expo`](../sdk/expo) package. There are a few ways to listen to events:

> - `useEvent` hook - this hook creates a listener that will return a stateful value that can be used in a component. It also cleans up automatically when the component unmounts.

Reviewer 2: ```suggestion
- `useEvent` hook: it creates a listener that will return a stateful value that can be used in a component. It also cleans up automatically when the component unmounts.
```

```tsx
import { useEvent } from 'expo';
// ... Other imports, definition of the component, creating the player etc.

const { status, error } = useEvent(player, 'statusChange', { status: player.status });
// Rest of the component...
```

> - `useEventListener` hook - built around the `Player.addListener` and `Player.removeListener` methods, creates an event listener with automatic cleanup.

Reviewer 2: ```suggestion
- `useEventListener` hook: built around the `Player.addListener` and `Player.removeListener` methods, creates an event listener with automatic cleanup.
```

```tsx
import { useEventListener } from 'expo';
// ...Other imports, definition of the component, creating the player etc.

useEventListener(player, 'statusChange', ({ status, error }) => {
  setPlayerStatus(status);
  setPlayerError(error);
  console.log('Player status changed: ', status);
});
// Rest of the component...
```

> - `Player.addListener` method - this is the most flexible way to listen to events, but requires manual cleanup and more boilerplate code.

Reviewer 2: ```suggestion
- `Player.addListener` method: this is the most flexible way to listen to events, but requires manual cleanup and more boilerplate code.
```
